File: app/src/main/java/com/dhpcs/liquidity/fragment/CreateGameDialogFragment.java
package com.dhpcs.liquidity.fragment;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDialogFragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;

import com.dhpcs.liquidity.R;
import com.dhpcs.liquidity.boardgame.BoardGame$;
import com.dhpcs.liquidity.ws.protocol.ZoneCommand$;

import java.util.ArrayList;
import java.util.Currency;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public class CreateGameDialogFragment extends AppCompatDialogFragment {

    public interface Listener {

        void onGameDetailsEntered(String name, Currency currency);

    }

    public static final String TAG = "create_game_dialog_fragment";

    private static class CurrenciesAdapter extends ArrayAdapter<Currency> {

        CurrenciesAdapter(Context context, List<Currency> currencies) {
            super(context, android.R.layout.simple_spinner_item, currencies);
            setDropDownViewResource(
                    android.R.layout.simple_spinner_dropdown_item
            );
            sort((lhs, rhs) -> lhs.getCurrencyCode().compareTo(rhs.getCurrencyCode()));
        }

        @TargetApi(Build.VERSION_CODES.KITKAT)
        private View bindView(TextView textView, Currency currency) {
            String displayName = Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT
                    ?
                    null
                    :
                    currency.getDisplayName();
            String symbolAndOrName;
            if (currency.getSymbol().equals(currency.getCurrencyCode())) {
                symbolAndOrName = displayName;
            } else {
                symbolAndOrName = displayName == null
                        ?
                        currency.getSymbol()
                        :
                        getContext()
                                .getString(
                                        R.string.game_currency_symbol_and_name_format_string,
                                        currency.getSymbol(),
                                        displayName
                                );
            }
            if (symbolAndOrName == null) {
                textView.setText(currency.getCurrencyCode());
            } else {
                textView.setText(
                        getContext().getString(
                                R.string.game_currency_format_string,
                                currency.getCurrencyCode(),
                                symbolAndOrName
                        )
                );
            }
            return textView;
        }

        @Override
        public View getDropDownView(int position, View convertView, @NonNull ViewGroup parent) {
            return bindView(
                    (TextView) super.getDropDownView(position, convertView, parent),
                    getItem(position)
            );
        }

        @NonNull
        @Override
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            return bindView(
                    (TextView) super.getView(position, convertView, parent),
                    getItem(position)
            );
        }

    }

    public static CreateGameDialogFragment newInstance() {
        return new CreateGameDialogFragment();
    }

    private Listener listener;

    private Button buttonPositive;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        listener = (Listener) context;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        @SuppressLint("InflateParams") View view = getActivity().getLayoutInflater().inflate(
                R.layout.fragment_create_game_dialog,
                null
        );

        Set<Currency> currencies = new HashSet<>();
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            for (Locale locale : Locale.getAvailableLocales()) {
                try {
                    currencies.add(Currency.getInstance(locale));
                } catch (IllegalArgumentException ignored) {
                }
            }
        } else {
            currencies.addAll(Currency.getAvailableCurrencies());
        }
        final ArrayAdapter<Currency> currenciesSpinnerAdapter = new CurrenciesAdapter(
                getActivity(),
                new ArrayList<>(currencies)
        );

        TextInputLayout textInputLayoutGameName = view.findViewById(R.id.textinputlayout_game_name);
        final TextInputEditText textInputEditTextGameName = view.findViewById(R.id.textinputedittext_game_name);
        final Spinner spinnerCurrency = view.findViewById(R.id.spinner_game_currency);

        final AlertDialog alertDialog = new AlertDialog.Builder(getActivity())
                .setTitle(
                        getString(
                                R.string.enter_game_details
                        )
                )
                .setView(view)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(
                        R.string.ok,
                        (dialog, whichButton) -> {
                            if (listener != null) {
                                listener.onGameDetailsEntered(
                                        textInputEditTextGameName.getText().toString(),
                                        currenciesSpinnerAdapter.getItem(
                                                spinnerCurrency.getSelectedItemPosition()
                                        )
                                );
                            }
                        }
                )
                .create();

        textInputLayoutGameName.setCounterMaxLength(ZoneCommand$.MODULE$.MaximumTagLength());
        textInputEditTextGameName.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (buttonPositive != null) {
                    validateInput(s);
                }
            }

        });
        spinnerCurrency.setAdapter(currenciesSpinnerAdapter);

        spinnerCurrency.setSelection(
                currenciesSpinnerAdapter.getPosition(
                        Currency.getInstance(Locale.getDefault())
                )
        );

        alertDialog.setOnShowListener(dialog -> {
            buttonPositive = alertDialog.getButton(DialogInterface.BUTTON_POSITIVE);
            validateInput(textInputEditTextGameName.getText());
        });

        Window window = alertDialog.getWindow();
        assert window != null;
        window.setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE
        );

        return alertDialog;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    private void validateInput(CharSequence gameName) {
        buttonPositive.setEnabled(
                BoardGame$.MODULE$.isGameNameValid(gameName)
        );
    }

}

File: app/src/main/java/com/dhpcs/liquidity/fragment/CreateGameDialogFragment.kt
package com.dhpcs.liquidity.fragment

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.findNavController
import com.dhpcs.liquidity.BoardGame
import com.dhpcs.liquidity.R
import com.dhpcs.liquidity.activity.MainActivity
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import java.util.*

class CreateGameDialogFragment : AppCompatDialogFragment() {

    companion object {

        private class CurrenciesAdapter internal constructor(context: Context,
                                                             currencies: List<Currency>
        ) : ArrayAdapter<Currency>(context,
                android.R.layout.simple_spinner_item,
                currencies) {

            init {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                sort { lhs, rhs -> lhs.currencyCode.compareTo(rhs.currencyCode) }
            }

            private fun bindView(textView: TextView, currency: Currency): View {
                val symbolAndOrName = if (currency.symbol == currency.currencyCode) {
                    currency.displayName
                } else {
                    context.getString(
                            R.string.game_currency_symbol_and_name_format_string,
                            currency.symbol,
                            currency.displayName
                    )
                }
                textView.text = context.getString(
                        R.string.game_currency_format_string,
                        currency.currencyCode,
                        symbolAndOrName
                )
                return textView
            }

            override fun getDropDownView(position: Int,
                                         convertView: View?,
                                         parent: ViewGroup
            ): View {
                return bindView(
                        super.getDropDownView(position, convertView, parent) as TextView,
                        getItem(position)!!
                )
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                return bindView(
                        super.getView(position, convertView, parent) as TextView,
                        getItem(position)!!
                )
            }

        }

        const val TAG = "create_game_dialog_fragment"

        fun newInstance(): CreateGameDialogFragment = CreateGameDialogFragment()

    }

    private var buttonPositive: Button? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        @SuppressLint("InflateParams") val view = requireActivity().layoutInflater.inflate(
                R.layout.fragment_create_game_dialog, null
        )

        val currenciesSpinnerAdapter = CurrenciesAdapter(
                requireContext(),
                ArrayList(Currency.getAvailableCurrencies())
        )

        val textInputLayoutGameName =
                view.findViewById<TextInputLayout>(R.id.textinputlayout_game_name)
        val textInputEditTextGameName =
                view.findViewById<TextInputEditText>(R.id.textinputedittext_game_name)
        val spinnerCurrency = view.findViewById<Spinner>(R.id.spinner_game_currency)

        val alertDialog = AlertDialog.Builder(requireContext())
                .setTitle(getString(
                        R.string.enter_game_details
                ))
                .setView(view)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.ok) { _, _ ->
                    val model = ViewModelProviders.of(requireActivity())
                            .get(MainActivity.Companion.BoardGameModel::class.java)
                    model.execCreateGameCommand(
                            model.boardGame.createGame(
                                    textInputEditTextGameName.text.toString(),
                                    currenciesSpinnerAdapter
                                            .getItem(spinnerCurrency.selectedItemPosition)!!,
                                    getString(R.string.bank_member_name)
                            )
                    ) {
                        getString(R.string.create_game_error)
                    }
                    findNavController().navigate(R.id.action_games_fragment_to_board_game_graph)
                }
                .create()

        textInputLayoutGameName.counterMaxLength = BoardGame.MAXIMUM_TAG_LENGTH
        textInputEditTextGameName.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) = validateInput(s)

        })
        spinnerCurrency.adapter = currenciesSpinnerAdapter

        spinnerCurrency.setSelection(
                currenciesSpinnerAdapter.getPosition(
                        Currency.getInstance(Locale.getDefault())
                )
        )

        alertDialog.setOnShowListener {
            buttonPositive = alertDialog.getButton(DialogInterface.BUTTON_POSITIVE)
            validateInput(textInputEditTextGameName.text!!)
        }

        val window = alertDialog.window!!
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)

        return alertDialog
    }

    private fun validateInput(gameName: CharSequence) {
        buttonPositive?.isEnabled = BoardGame.isGameNameValid(gameName)
    }

}
